### ISSUE TITLE
Add function signatures / declarations...

### ISSUE THREAD

**Issue by nihilus** [yegord/snowman#16]
With some effort this could be implemented based on van Emmerik's approach in boomerang. Allowing customization of header files with function declarations in-place. It would improve the readability and the typechecking of decompiled code.

**Comment by yegord** [comment]
I am not sure if I get the point of the task: is this about the recognition of standard functions and using the knowledge about their signatures?

**Comment by nihilus** [comment]
The recognition of "standard" functions like IDA does it. Then knowing the arguments and the type of the arguments one is able to propagate the type thru the decompiled program.

**Comment by nihilus** [comment]
Actually the C++ demangling seems to give the types of the arguments which could be piggy-backed on and propagated. :+1:

**Comment by jbvsmo** [comment]
Hi, I didn't want to open a new issue and this one seems about right. 
I do have a header file and a shared lib which does not decompile very well as it has a huge amount of structs with the same format and incremental names: `struct s1234 { ... };`

I believe a much cleaner code could be produced from extracting this information from the header file and have actual structs access show the members names instead of bunch of `ecx->0x8 = a9` 

it would be: `element->variance = calc_variance` since ecx and a9 are from the function signature and the variance from struct.

Also even without the header, a bunch more info is available such as this class constructor example:

```
/*
 * ??0FnName@FnClass@my_namespace@@AAA@_xxxxxBlaBlaBla@14@AAAA@Z
 * protected: __thiscall my_namespace::FnClass::0FnName::0FnName(bool, bool, bool, bool, bool, bool, enum my_namespace::FnClass::OtherClass, unsigned int, unsigned int, double, double)
 */
struct s1273* _0FnName_FnClass_my_namespace_AAA__xxxxxBlaBlaBla_14_AAAA_Z(struct s1823* ecx, signed char a2, signed char a3, signed char a4, signed char a5, signed char a6, signed char a7, int32_t a8, int32_t a9, int32_t a10) {
```

All the real types are annotated there above and with the header file we would have the variable names as well.

I am willing to help with this if possible

**Comment by yegord** [comment]
This all makes sense. However, to get this feature, somebody has to essentially integrate a compiler frontend into Snowman. I doubt somebody capable of doing it is really going to do spend his/her free time on it.

**Comment by jbvsmo** [comment]
Can't this e.g. optionally integrate with clang and just get an AST or something?

`clang -x c++ -Xclang -ast-dump -fsyntax-only foo.h`

**Comment by yegord** [comment]
Yes, why not.
If you can take types out of clang, with their sizes, in case of structs — with their fields and offsets, one could use that.
It would be nice to know calling conventions of functions, and details about which parameter goes where.
But I do not readily see anything from this list in the AST dump.

**Comment by jbvsmo** [comment]
Calling convention should depend on compiler/OS from original binary build. Clang cannot give you that information.

But I believe you can better guess once you know how many parameters there are and their actual defined types.

I realize the current code does miss some function parameters now and then as in the example above where the `double` elements at the end do not exist on the reconstituted function.

Also, the 1st parameter is `this` pointer which can be inferred to exist because you would know it is a method of a class (in that case the constructor).

**Comment by yegord** [comment]
One way or another, the decompiler will need calling conventions of functions declared in .h, layouts of structs to meaningfully apply the type information from there to abstract memory locations (e.g., stack pointer offsets). One way to get them is to ask a compiler. Another is to write this (part of) compiler yourself. Doing any of the two is a big chunk of work already.

**Comment by jbvsmo** [comment]
Maybe I misunderstood what you're trying to say, but don't you assume one particular calling convention when transforming asm into function calls on the C representation?

E.g. Win32 you could see the symbol name and decide how it is being called: http://unixwiz.net/techtips/win32-callconv.html

So lets say you assume a particular function is cdecl on the binary and you know from the header file or from the symbol annotation that it has 3 parameters `bool, int, long long` then you can assume the stack is 1 + 4 + 8 bytes high on the caller to generate `foo(a, b, c);`. If the header also contains the parameters names, you could propagate back to make caller code readable as people tend to use same names.

So I don't understand what else would be missing in terms of calling convention. BTW, I said I can try to help unless you're not interested and having such feature.

**Comment by yegord** [comment]
> don't you assume one particular calling convention when transforming asm into function calls on the C representation?

The decompiler guesses the calling convention for each function.

> Win32 you could see the symbol name and decide how it is being called

The decompiler does exactly this when guessing calling conventions for a Win32 binary.

> it has 3 parameters bool, int, long long then you can assume the stack is 1 + 4 + 8 bytes high on the caller to generate foo(a, b, c);

You also need to know that bool is 1 byte long, int is 4 bytes long, long is 8 bytes long.
You also need to know the alignment rules: actually, the last argument will be not 13 bytes high on the stack, but some round number like 16 bytes high.
When a struct is passed, you must know the size of the struct, the offsets of all its fields.
Sometimes rules of passing structs are pretty weird: structs smaller that a given size are passed in a couple of registers, and bigger structs are passed on the stack.

The decompiler must know all that before it can match the abstract memory locations (and their types) with the arguments from the header file.So, you need to have this logic somewhere: either in the compiler, and it must say it to you what is what and passed where, or in the decompiler.

Plus, you need to reimplement the compiler's mangling, in order to be able to match functions in the header files with the symbols in the executable (or again ask the compiler).

To mangle properly, you again need all the type information, i.e., all typedefs used, all dependencies of the given header file.

So, every item is a bit of work. If you are willing to do this work for free and produce industry-level code, with tests, compliant with the coding style, I'll do my best to find some time and energy to review it.

**Comment by jbvsmo** [comment]
> The decompiler guesses the calling convention for each function.

That is what I meant. When you pick one you know which registers receive what and where each thing is on the stack, possibly. You keep saying you need to ask the compiler, but you need to ask the *original* compiler and not the one you will theoretically be using for this purpose. Therefore of course you need to setup the rules like struct return on Microsoft cdecl taking 1 or 2 registers if small enough. You need to know the original size of the types like `long`, `bool` used by the *original* compiler and so on. Same with memory alignment. You can extract all this information from the binary itself which I know is some work, but extremely doable.

So that means I do have a bit of time to invest and this tool did help me a few times in the past that I don't mind giving some back. If it was implied anywhere, I seek zero money whatsoever and would be just as offended if someone offered it.

But if you need something rigorously perfect before you can even look into it, I am better off half-assing a python script tied to a compiler command line and regex the hell of that to get me stuff I need for a project I'm doing.

**Comment by yegord** [comment]
> but you need to ask the original compiler and not the one you will theoretically be using for this purpose

Exactly. However, generally you cannot ask the original compiler. You will have to ask _a_ compiler that can parse an .h file for some of the information and infer the rest in the decompiler itself. And deciding what to get from a compiler and what to extract in the decompiler itself is another piece of work to be done: the borderline is not immediately clear to me now.

> But if you need something rigorously perfect before you can even look into it

Maybe not rigorously perfect, but without obvious bugs at least. I already do enough code reviews at work in order to not to seek for another painful code reviews in my free time. :)

**Comment by brandonros** [comment]
![image](https://user-images.githubusercontent.com/8949910/171938943-c320be7e-ae9c-4021-a0a1-953204f1c4cb.png)

Is there a way to override what Snowman comes up with when it is like... wrong?

```c
typedef T_PDU_ERROR (*T_PDUConstruct)(CHAR8* OptionStr, void *pAPITag);
```

That's a commonly known DLL export for a public API that is only 2 arguments, but as you can see Snowman detects 3. Is there a way to tell Snowman like... it's wrong? It's only a 2 argument function?

**Comment by yegord** [comment]
> Is there a way to override what Snowman comes up with when it is like... wrong?

No, this was never implemented.
